### ISSUE TITLE
Keras model save/load is totally wrong with sessions

### ISSUE THREAD

**Issue by wookayin** [tensorflow/tensorflow#26472]
## Bug report: Keras `model.save()` does not work with sessions

Actually, it only works with the default session.

I know tf.keras is mainly targetting at TF 2.0, but it really doesn't play nice with conventional TF-1.x ways (e.g. Session). I encountered this bug and nailed down why it happens after a day. Arguably, I believe this is a severe bug and hard to realize when one encounters. This is also somewhat related to #26430.

**In summary (the current behavior)**: if one is using a custom session created (say `MonitoredSession`, as it was a conventional way of opening and sessions in TF 1.x way), Keras model's load/save will be COMPLETELY not working.

### System information

- OS Platform and Distribution: Linux, but agnostic to Platform
- TensorFlow installed from (source or binary): binary, pypi
- TensorFlow version (use command below): **1.13.1**
- Python version: 3.6

## Code to reproduce the issue

Please see this notebook:

https://gist.github.com/wookayin/0712261896799e0f655df50e910055df

Or in a code:

```python
import tensorflow as tf

class MyModel(tf.keras.models.Model):
    def __init__(self):
        super(MyModel, self).__init__()
        self.layer = tf.keras.layers.Dense(5)
        
    def __call__(self, x):
        return self.layer(x)
    
model = MyModel()

# Build the model
model(tf.zeros([3, 10]))

# this is something like a train_op that "changes" the content of the variable.
assign_op = model.variables[1].assign( tf.ones_like(model.variables[1]) )

# Let's create a session
sess = tf.train.SingularMonitoredSession()
#sess.run(tf.global_variables_initializer())

# we can see the 'bias' variable is initialized to ZERO
assert sess.run(model.variables[1]).mean() == 0.0

# Now let's make the 'bias' variable to all one...
sess.run(assign_op)

# sure it is ONE ...
assert sess.run(model.variables[1]).mean() == 1.0

# Let's save the Keras model parameter. The bias is set to ONE, right ?????
# Since model.save_weights try to create a new op (another bug #26430)
# and the graph has been finalized, we will 'unfinalize' the graph with a bit of hack
sess.graph._finalized = False
model.save_weights('/tmp/keras-one.h5')
sess.graph.finalize()


# Let's see what is stored in the model file ....
import h5py
h = h5py.File("/tmp/keras-one.h5")
assert h['dense']['dense']['bias:0'].value.mean() == 1.0     # <------ This will fail
# Actual output is: array([0., 0., 0., 0., 0.], dtype=float32)
```

The above code does the following:

* Build a Keras model
* Initialize the layer (bias will be zero)
* Update the trainable variable (e.g. bias: 0->1)
* Save the model into a H5 File
* Read the model file, but the bias variable is not 1 but 0 (same as the initialization)

### Additional Information

* `tf.Session()` behaves exactly the same as with `tf.train.MonitoredSession()`. In this case we need to run custom variable initializer ops.
* However, when `tf.InteractiveSession()` is used, it works exactly as expected (the readout from `keras-one.h5` is 1.0)
* It would work well in the eager mode.

## A work-around

If we change the code as follows, it would work.

```
with sess.as_default():
    model.save_weights('/tmp/keras-one.h5')
```

A quick note: 'SingularMonitoredSession' object has no attribute 'as_default'

This is because [Keras only plays with **the default session**][keras-session] (see #26430 as well) and in this example the manually created session was never registered as default. Kinda makes sense as there is no way for `model.save()` function to be aware of the session... However, this leads to a **very inconsistent and unexpected behavior.** If no session is registered as default, `model.save()` should instead throw an error.

[keras-session]: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/backend.py#L401

**Comment by wookayin** [comment]
Any feedbacks? I think we should make tf.Keras **NOT** use its own default session (`_SESSION`) when integrated to TF core. However, it can also break many things that have been relying on this implicit Keras session.

**Comment by zh794390558** [comment]
This is an most important problem. Please check it .

**Comment by wookayin** [comment]
@fchollet @ymodak Can we revisit this issue? As TF 1.x is expecting the last release soon, I hope this bug should be fixed.

**Comment by DavidS3141** [comment]
If this is not fixed quite soon, one should at least introduce a warning and some documentation about this behavior, this was quite unexpected for me and wasted a few trainings.

**Comment by aashnajena** [comment]
Has this been solved for keras 2.x? I'm using 2.3.1 and I can still see this problem.
I'm using Jupyter notebooks. In one notebook, I'm saving a model using `model.save()` and later loading the model using `load_model()` and it works perfectly fine. 
But when I try to open the same saved model from a different notebook, I get very random accuracy and predictions. From what I understand, the weights are getting reinitialized randomly. What is the workaround for this? Urgent.

**Comment by wookayin** [comment]
@aashnajena Make sure the session you are using is registered as a "default" session. I already have put a code snippet for the workaround.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by wookayin** [comment]
This still needs to be fixed in 2.x.

**Comment by rmothukuru** [comment]
@wookayin,
Can someone provide a reproducible code in `Tensorflow 2.x` so that it can be looked into? Thanks!
